PagingWithNetworkSample/app/src/main/java/com/android/example/paging/pagingwithnetwork/reddit/repository/inMemory/byPage/PageKeyedSubredditPagingSource.kt

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (C) 2017 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.example.paging.pagingwithnetwork.reddit.repository.inMemory.byPage
+
+import androidx.paging.LoadType.*
+import androidx.paging.PagingSource
+import androidx.paging.PagingSource.LoadResult.Page
+import com.android.example.paging.pagingwithnetwork.reddit.api.RedditApi
+import com.android.example.paging.pagingwithnetwork.reddit.vo.RedditPost
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.withContext
+import java.io.IOException
+
+/**
+ * A data source that uses the before/after keys returned in page requests.
+ *
+ * @see [com.android.example.paging.pagingwithnetwork.reddit.repository.inMemory.byItem.ItemKeyedSubredditPagingSource]
+ */
+class PageKeyedSubredditPagingSource(
```

Comment:
Unify all this stuff, by using just one Retrofit call, e.g.:

    @GET("/r/{subreddit}/hot.json")
    suspend fun getTop(
            @Path("subreddit") subreddit: String,
            @Query("limit") limit: Int,
            @Query("after") after: String? = null,
            @Query("before") before: String? = null
    ): ListingResponse
 
Which lets us massively simplify the entire PagingSource to something like:
```
class PageKeyedSubredditPagingSource(
        private val redditApi: RedditApi,
        private val subredditName: String,
        private val networkDispatcher: CoroutineDispatcher
) : PagingSource<String, RedditPost>() {
 
    override suspend fun load(params: LoadParams<String>): LoadResult<String, RedditPost> {
        return try {
            val data = withContext(networkDispatcher) {
                redditApi.getTopAfter(
                        subreddit = subredditName,
                        after = if (params.loadType == END) params.key else null,
                        before = if (params.loadType == START) params.key else null,
                        limit = params.loadSize
                ).data
            }
            Page(
                    data = data.children.map { it.data },
                    prevKey = data.before,
                    nextKey = data.after
            )
        } catch (e: IOException) {
            LoadResult.Error(e)
        }
    }
}
```

Reply:
Done